### PR TITLE
Make OnConnectionStateChange wait for handler to finish

### PR DESCRIPTION
**Pull request by sukunrt**
`OnConnectionStateChange` is used to determine whether the peer connection has succeeded and track connection state by users. It should be called in sync. Right now state changes might get reordered which would lead to tracking errors for users.

### CODE REVIEW

File: peerconnection.go [hunk 1]
```diff
@@ -510,7 +510,13 @@ func (pc *PeerConnection) onConnectionStateChange(cs PeerConnectionState) {
...
 	pc.log.Infof("peer connection state changed: %s", cs)
 	if handler, ok := pc.onConnectionStateChangeHandler.Load().(func(PeerConnectionState)); ok && handler != nil {
-		go handler(cs)
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			handler(cs)
+			wg.Done()
+		}()
+		wg.Wait()
```
Comment (Wondertan) [inline peerconnection.go:519]: Why wg instead of just removing `go`?

Comment (Sean-Der) [reply peerconnection.go:519]: @Wondertan this is so the callback doesn’t block the ICE Agent

hopefully we have a test for that :)

Comment (Sean-Der) [reply peerconnection.go:519]: I believe we have this issue with this PR. The code right now would allow a callback to block the PeerConnection internals.

**Review by Sean-Der — CHANGES_REQUESTED** [review]

File: peerconnection.go (hunk 1 above)
Comment (sukunrt) [reply peerconnection.go:519]: >Why wg instead of just removing go?

Yeah sorry this is not required at all.

**Comment by sukunrt** [comment]
I have moved the updates to a operations queue. That'll run the notifications sequentially. This ensures that users can do things like `pc.Close` from the notification.

File: peerconnection.go (hunk 1 above)
Comment (Wondertan) [reply peerconnection.go]: No worries @sukunrt, all good

**Comment by Sean-Der** [comment]
This was fixed by https://github.com/pion/ice/commit/67cc918a518d3b3dd1887b07434331dabc846702, going to close here!

Will be tagging a new version of pion/ice today! thank you @sukunrt :)

**Comment by sukunrt** [comment]
It's not exactly fixed. The state reported on the callback isn't reliable. Users need to ignore the state provided on the call back and call `pc.ConnectionState()` for the accurate state of the connection.

**Pull request closed without merging**